sql: Alter Column Type Using Expression

RichardJCai opened the pull request:
sql: support alter column type using expression

Fixes #47706

Release note (sql change): Support ALTER COLUMN TYPE ... USING EXPRESSION
to allow users to optionally provide an expression to convert a column to a new type.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49452)
<!-- Reviewable:end -->

RichardJCai commented:
Would like some comments on the use of crdb_internal.force_error() in the old column's computed expression to disable writes into the old column.

rohany commented on pkg/sql/sem/tree/alter_table.go:
```diff
@@ -219,6 +219,7 @@ type AlterTableAlterColumnType struct {
 	Column    Name
 	ToType    ResolvableTypeReference
 	Using     Expr
+	Inverse   Expr
```
I think this is unused

rohany replied:
bumping

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3272,13 +3272,33 @@ func (desc *MutableTableDescriptor) performComputedColumnSwap(swap *ComputedColu
+			"Column %s is undergoing the ALTER COLUMN TYPE USING EXPRESSION "+
+				"schema change. Inserts are not supported until the schema change is "+
+				"finalized", oldCol.Name)
+		insertedValToString := tree.CastExpr{
```
This is going to fail with a confusing error if the old expression type can't be converted into a string right (do we have types that can't be converted into strings?).

RichardJCai replied:
It doesn't look like there are any types that can't be converted to STRING currently. I wonder if this might be a problem in the future.

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3272,13 +3272,33 @@ func (desc *MutableTableDescriptor) performComputedColumnSwap(swap *ComputedColu
+			SyntaxMode: tree.CastShort,
+		}
+		insertedVal := tree.Serialize(&insertedValToString)
+		s := fmt.Sprintf(
```
I think it would be helpful for readability to include a comment here about what the computed expression will look like, as well as why this is safe to do.

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -442,3 +435,39 @@ CREATE TABLE t22 (x INT);
+INSERT INTO t23 VALUES (-3), (-2), (-1), (0), (1), (2), (3)
+
+statement ok
+ALTER TABLE t23 ALTER COLUMN x TYPE BOOL USING (x > 0)
```
What happens if the using expression doesn't agree with the new type of the column? What if there are invalid variable references? Are there certain classes of functions that we allow in the using expression? It would probably be similar to what we allow in computed columns.

RichardJCai replied:
We do that to disable invalid variable references but I'm not sure if we would have the same set of disallowed functions as computed columns. Postgres seems to allow impure functions for the expression - not sure if we would want to support that.

rohany commented on pkg/sql/sqlbase/structured.proto:
```diff
@@ -523,6 +523,8 @@ message ComputedColumnSwap {
   optional uint32 new_column_id = 1 [(gogoproto.nullable) = false, (gogoproto.casttype) = "ColumnID"];
   optional uint32 old_column_id = 2 [(gogoproto.nullable) = false, (gogoproto.casttype) = "ColumnID"];
+  // Used to indicate that ALTER COLUMN TYPE ... USING EXPR was used.
+  optional bool using_expr = 3 [(gogoproto.nullable) = false];
```
Would it be helpful to send over the "inverse" expr here, rather than a bool? In the future if we support inverse, we'll send that over, otherwise we'll send over this always error expression.

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3272,13 +3272,46 @@ func (desc *MutableTableDescriptor) performComputedColumnSwap(swap *ComputedColu
+	// the EXPRESSION to insert into the old column and thus will force an
+	// error using the computed expression. Any inserts into the new column
+	// will fail until the old column is dropped.
+	if swap.InverseExpr != nil && *swap.InverseExpr == tree.Serialize(tree.DNull) {
```
I meant that always serialize this erroring cast during planning and send it over in `InverseExpr`. If inverse expr != nil, then use it, otherwise create a basic cast here. Is that not possible?

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3272,13 +3272,46 @@ func (desc *MutableTableDescriptor) performComputedColumnSwap(swap *ComputedColu
+		// This is safe to do so because after the column swap, the old column
+		// should become "read-only", new inserts should only be possible via
+		// the computed expression.
+		// An example of the error:
```
I meant a string of the computed expression -- the error message already printed above.

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -442,3 +435,56 @@ CREATE TABLE t22 (x INT);
+statement error pq: no data source matches prefix: t26
+ALTER TABLE t25 ALTER COLUMN x TYPE BOOL USING (t26.x > 0);
+
+# Ensure USING EXPRESSION rolls back if the USING EXPRESSION does not conform to the new type of the column.
```
[nit] line length

RichardJCai replied:
I haven't really been watching line lengths in logic tests since a lot of the errors are past the line length limit.

rohany replied:
It's good to keep the comments around 80 characters for readability still, even if the errors go larger than the limit.

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -231,14 +222,75 @@ func alterColumnTypeGeneral(
+	if Using != nil {
+		// Ensure the USING EXPRESSION returns the new type.
+		// Replace column references with typed dummies to allow typechecking.
+		replacedExpr, _, err := replaceVars(tableDesc, Using)
```
I've seen this logic in a lot of places now -- is there something in the `schemaexpr` package that does this for you?

RichardJCai replied:
I created a helper function to do this replace and sanitize that seems to be a reoccuring pattern.

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -231,14 +222,75 @@ func alterColumnTypeGeneral(
+			return err
+		}
+
+		_, err = sqlbase.SanitizeVarFreeExpr(
```
You'll want to serialize this typed expression so that it works with UDTs (see  the changes in #49565 )

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -231,14 +222,75 @@ func alterColumnTypeGeneral(
+	} else {
+		// The default computed expression is casting the column to the new type.
+		newComputedExpr := tree.CastExpr{
+			Expr:       &tree.ColumnItem{ColumnName: tree.Name(col.Name)},
```
Does anything weird happen here by setting the name of the column in the expression here -- I know we pick new names for columns in the mutation.

RichardJCai replied:
I think this is the correct behaviour since both the computed expressions should be based on the name of the current column. When backfilling the old column, we use the current column and then the "inverse" expression is used after the swap so the new column will now be named as the original column name.

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3269,16 +3269,8 @@ func (desc *MutableTableDescriptor) performComputedColumnSwap(swap *ComputedColu
-	}
-	s := tree.Serialize(&oldColComputeExpr)
-	oldCol.ComputeExpr = &s
+	// Make the oldCol a computed column by setting it's computed expression.
```
[nit] its

rohany commented on pkg/sql/sqlbase/structured.proto:
```diff
@@ -523,6 +523,11 @@ message ComputedColumnSwap {
   option (gogoproto.equal) = true;
   optional uint32 new_column_id = 1 [(gogoproto.nullable) = false, (gogoproto.casttype) = "ColumnID"];
   optional uint32 old_column_id = 2 [(gogoproto.nullable) = false, (gogoproto.casttype) = "ColumnID"];
+  // Currently this field will always be NULL and is used to check if a column
```
This comment is out of date now right?

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -231,14 +222,75 @@ func alterColumnTypeGeneral(
+	// To achieve this, we make oldCol a computed column of newCol using
+	// inverseExpr.
+	// If a USING EXPRESSION was provided, we cannot generally invert
+	// the EXPRESSION to insert into the old column and thus will force an
```
[nit] shoudl expression be all caps here

rohany reviewed (commented):
LGTM with some nits. I'd like for @jordanlewis to take a peek at this use of the error builtin

RichardJCai commented:
I also fixed a small bug with the replaceVars logic, it didn't explicitly forbid column references from other tables when checking expressions. Because of this, there's a bug on master where you can refer to columns in other tables when creating computed columns, check exprs, etc.
The column will get normalized and it will refer to the column in the current table. Heres an example.

```
root@127.0.0.1:54117/movr> create table t(x int, y int);
CREATE TABLE

Time: 2.383ms

root@127.0.0.1:54117/movr> create table t2(x int);
CREATE TABLE

Time: 3.252ms

root@127.0.0.1:54117/movr> insert into t2 values (1);
INSERT 1

Time: 2.512ms

root@127.0.0.1:54117/movr> alter table t add column z int as (t2.x + 1) stored;
root@127.0.0.1:54117/movr> show create table t;
  table_name |             create_statement
-------------+--------------------------------------------
  t          | CREATE TABLE t (
             |     x INT8 NULL,
             |     y INT8 NULL,
             |     z INT8 NULL AS (x + 1:::INT8) STORED,
             |     FAMILY "primary" (x, y, rowid, z)
             | )
```

rohany commented:
I'm surprised that this bug was on master -- nice find! Did it basically just check if i had `<othertable>.col` that `col` was a valid column in the target table?

RichardJCai commented:
> I'm surprised that this bug was on master -- nice find! Did it basically just check if i had `<othertable>.col` that `col` was a valid column in the target table?

Yeah that's exactly it.

solongordon reviewed (changes requested):
Agree the `force_error` business feels hacky but I don't know that it's a dealbreaker if it's providing the behavior we want. 

> Reviewed 2 of 11 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-:-M9PB0hq0tdN0DUB46Gf:b-1ss9gk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, and @rohany)

___
*[pkg/sql/alter_column_type.go, line 231 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0Ef9V4B7KoTu-rvr-r3-231:-M9P64CZ9JYpUT733eRG:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f413fe01b2b91a8abc3720ddd2fd347f38658f/pkg/sql/alter_column_type.go#L231)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

[nit] shoudl expression be all caps here
</blockquote></details>

+1
___
*[pkg/sql/alter_column_type.go, line 265 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9Ox4gW7AhlfJpv9kuU:-M9Ox4gW7AhlfJpv9kuV:bt36u47) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/alter_column_type.go#L265)):*
> ```Go
> 		// crdb_internal.force_error() will cause the write to error out.
> 		// This prevents the column going ALTER COLUMN TYPE to be unable to write
> 		// to until the old column is dropped.
> ```

This sentence doesn't make sense to me.
___
*[pkg/sql/alter_column_type.go, line 276 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9P6w3k6WpCzxB33unG:-M9P6w3k6WpCzxB33unH:b4d58h1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/alter_column_type.go#L276)):*
> ```Go
> 		failedInsertMsg := fmt.Sprintf(
> 			"'\nTried to insert ', %s, ' into %s'", insertedVal, col.Name,
> 		)
> ```

I think you could omit this line.
___
*[pkg/sql/alter_column_type.go, line 278 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9P70Wj0sIy_RJ9R7Ue:-M9P70Wj0sIy_RJ9R7Uf:bny8vjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/alter_column_type.go#L278)):*
> ```Go
> 		)
> 		inverseExpr = fmt.Sprintf(
> 			"crdb_internal.force_error('03000', concat(%s, %s))",
> ```

Could you parametrize '03000' here so it's not just a magic number? Also the use of `concat` here is kind of funky. I'd say do the string concatenation on the Go side.
___
*[pkg/sql/logictest/testdata/logic_test/alter_column_type, line 479 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9OuqOH2BSJ00vKzoFi:-M9OuqOH2BSJ00vKzoFj:b6pjzv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/logictest/testdata/logic_test/alter_column_type#L479)):*
> ```
> )
> 
> # Ensure USING EXPRESSION cannot reference other columns.
> ```

Do you mean other tables? Referencing columns from the same table seems allowed.
___
*[pkg/sql/schemaexpr/column.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9PACcOC_BWAHLQxvNI:-M9PACcOC_BWAHLQxvNJ:bekx0ty) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/schemaexpr/column.go#L243)):*
> ```Go
> 		if c.TableName != nil {
> 			name := c.TableName.ToTableName()
> 			if name.Table() != desc.Name {
> ```

This would still allow referencing a table with the same name in a different database.
___
*[pkg/sql/schemaexpr/column.go, line 246 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9PAT6P3fsCsXW2Wguw:-M9PAT6P3fsCsXW2Wgux:byh7kah) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/schemaexpr/column.go#L246)):*
> ```Go
> 				return false, nil,
> 					errors.Newf("cannot reference columns from other tables\n"+
> 						"tried to reference column from table %s", name.String())
> ```

Could you use `pgerror.Newf` with an appropriate `pgcode`? Also having a newline in an error msg seems weird to me.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, and @rohany)

___
*[pkg/sql/alter_column_type.go, line 231 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0Ef9V4B7KoTu-rvr-r3-231:-M9PFsAGE7q0tIVUxZPk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f413fe01b2b91a8abc3720ddd2fd347f38658f/pkg/sql/alter_column_type.go#L231)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.
___
*[pkg/sql/alter_column_type.go, line 238 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0Ef9V4B7KoTu-rvr-r3-238:-M9PFt4q1Uf4O7HXRKQL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f413fe01b2b91a8abc3720ddd2fd347f38658f/pkg/sql/alter_column_type.go#L238)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I created a helper function to do this replace and sanitize that seems to be a reoccuring pattern.
</blockquote></details>

Done.
___
*[pkg/sql/alter_column_type.go, line 243 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0Ef9V4B7KoTu-rvr-r3-243:-M9PFtau7fhOtoP-gn-s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f413fe01b2b91a8abc3720ddd2fd347f38658f/pkg/sql/alter_column_type.go#L243)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

You'll want to serialize this typed expression so that it works with UDTs (see  the changes in #49565 )
</blockquote></details>

Done.
___
*[pkg/sql/alter_column_type.go, line 280 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0Ef9V4B7KoTu-rvr-r3-280:-M9PFubn5NHCjQ1nX65S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f413fe01b2b91a8abc3720ddd2fd347f38658f/pkg/sql/alter_column_type.go#L280)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I think this is the correct behaviour since both the computed expressions should be based on the name of the current column. When backfilling the old column, we use the current column and then the "inverse" expression is used after the swap so the new column will now be named as the original column name.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/alter_column_type, line 445 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0EfArai3e557k0Au-r1-445:-M9PFv5NAEWm0te2l65v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7147f659f05004b1a3846aede1c7f030d4c42e2d/pkg/sql/logictest/testdata/logic_test/alter_column_type#L445)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

We do that to disable invalid variable references but I'm not sure if we would have the same set of disallowed functions as computed columns. Postgres seems to allow impure functions for the expression - not sure if we would want to support that.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/alter_table.go, line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0EfBFghzAxuTeGRh-r1-222:-M9PFle6ANqB8mhPuNZW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7147f659f05004b1a3846aede1c7f030d4c42e2d/pkg/sql/sem/tree/alter_table.go#L222)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

bumping
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.go, line 3284 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0EfArai3e557k0B0-r1-3284:-M9PFn7VAIvKUJ7FkfB1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7147f659f05004b1a3846aede1c7f030d4c42e2d/pkg/sql/sqlbase/structured.go#L3284)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

It doesn't look like there are any types that can't be converted to STRING currently. I wonder if this might be a problem in the future.
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.go, line 3290 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0EfArai3e557k0B0-r1-3290:-M9PFnuo8MkU6L-xT08Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7147f659f05004b1a3846aede1c7f030d4c42e2d/pkg/sql/sqlbase/structured.go#L3290)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I think it would be helpful for readability to include a comment here about what the computed expression will look like, as well as why this is safe to do.
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.go, line 3279 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0EfArai3e557k0B0-r2-3279:-M9PFoIwD0DLp76VeCHN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/020ba6a6ef37a6db94eefa1502a813eeb1ba40f2/pkg/sql/sqlbase/structured.go#L3279)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I meant that always serialize this erroring cast during planning and send it over in `InverseExpr`. If inverse expr != nil, then use it, otherwise create a basic cast here. Is that not possible?
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.go, line 3298 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0EfArai3e557k0B0-r2-3298:-M9PFoer3jGsrbsPtMpp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/020ba6a6ef37a6db94eefa1502a813eeb1ba40f2/pkg/sql/sqlbase/structured.go#L3298)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I meant a string of the computed expression -- the error message already printed above.
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.go, line 3272 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0EfArai3e557k0B0-r3-3272:-M9PFmc497Ouf8xX4wkW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f413fe01b2b91a8abc3720ddd2fd347f38658f/pkg/sql/sqlbase/structured.go#L3272)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

[nit] its
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.proto, line 527 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0EfArai3e557k0B2-r1-527:-M9PFpWU0kdssyI-zgjj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7147f659f05004b1a3846aede1c7f030d4c42e2d/pkg/sql/sqlbase/structured.proto#L527)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Would it be helpful to send over the "inverse" expr here, rather than a bool? In the future if we support inverse, we'll send that over, otherwise we'll send over this always error expression.
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.proto, line 526 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0EfArai3e557k0B2-r3-526:-M9PFpDe0mMSBIsog5rw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f413fe01b2b91a8abc3720ddd2fd347f38658f/pkg/sql/sqlbase/structured.proto#L526)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

This comment is out of date now right?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, @rohany, and @solongordon)

___
*[pkg/sql/alter_column_type.go, line 276 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9P6w3k6WpCzxB33unG:-M9PHWM7-EIBsLupQuZI:b-qla35k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/alter_column_type.go#L276)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I think you could omit this line.
</blockquote></details>

This line is used to actually trigger the computed expression when a value is inserted into the new table. We need the computed expression to depend on the new column, I couldn't really think of a better way to do this besides using this sort of message.
___
*[pkg/sql/logictest/testdata/logic_test/alter_column_type, line 479 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9OuqOH2BSJ00vKzoFi:-M9PFy7z0PCsC8ReKt3H:b8cdsbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/logictest/testdata/logic_test/alter_column_type#L479)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Do you mean other tables? Referencing columns from the same table seems allowed.
</blockquote></details>

Oops, yeah I meant table.
___
*[pkg/sql/schemaexpr/column.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9PACcOC_BWAHLQxvNI:-M9PG0vo3_0NeFnLmJGO:b-a22chf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/schemaexpr/column.go#L243)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

This would still allow referencing a table with the same name in a different database.
</blockquote></details>

You're right, whats tricky about this is that we don't actually store the database name in the ColumnItem at the moment, so if reference a table with the same name but in another database, it would work.

I think this issue is fairly tedious to solve because we have to go and add database name into ColumnItem in every place a ColumnItem is needed, I just checked and a lot of code that creates a new column doesn't have database name handy to pass in.

Is it alright if I file a separate issue for this?


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, @rohany, and @solongordon)

___
*[pkg/sql/schemaexpr/column.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9PACcOC_BWAHLQxvNI:-M9P_bnhDwA0UGwJmsXr:bh01npd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/schemaexpr/column.go#L243)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

You're right, whats tricky about this is that we don't actually store the database name in the ColumnItem at the moment, so if reference a table with the same name but in another database, it would work.

I think this issue is fairly tedious to solve because we have to go and add database name into ColumnItem in every place a ColumnItem is needed, I just checked and a lot of code that creates a new column doesn't have database name handy to pass in.

Is it alright if I file a separate issue for this?
</blockquote></details>

ColumnItems contain a `UnresolvedObjectName`, which you can access the database, schema, and table from


<!-- Sent from Reviewable.io -->

RichardJCai commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, @rohany, and @solongordon)
> 
> _[pkg/sql/schemaexpr/column.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9PACcOC_BWAHLQxvNI:-M9P_bnhDwA0UGwJmsXr:bh01npd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/schemaexpr/column.go#L243)):_
> 
> Previously, RichardJCai (Richard Cai) wrote…
> ColumnItems contain a `UnresolvedObjectName`, which you can access the database, schema, and table from

Wait yeah you're right it I thought it only contained the actual name of the table at first glance.

RichardJCai commented:
I wonder if it would be alright to disable the use of database references (and schema? probably not for schema even thought we don't really support user defined schemas right now) identified table names completely.

We don't have the database/schema name handy when we do this check, would have to go and retrieve the DatabaseDescriptor using the TableDescriptors ID.

The one caveat being Postgres does support providing these identifiers for views but not for check constraints or computed columns.

You get this error when trying to reference cross database columns in postgres
```
d=# create table t3(x int generated always as (postgres.t.x > 0) stored);
ERROR:  missing FROM-clause entry for table "t"
LINE 1: create table t3(x int generated always as (postgres.t.x > 0)...
```

@rohany @solongordon

rohany commented:
That seems reasonable to me -- If the unresolved name has > 1 part, then it's going to be possibly referencing not the current table, so i think its fine to just block that use.

RichardJCai commented:
> That seems reasonable to me -- If the unresolved name has > 1 part, then it's going to be possibly referencing not the current table, so i think its fine to just block that use.

Pushed a patch to disallow specifying columns and db names when referring to columns.

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rohany, and @solongordon)

___
*[pkg/sql/alter_column_type.go, line 265 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9Ox4gW7AhlfJpv9kuU:-M9Q6Onc944SOH2Wr4yG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/alter_column_type.go#L265)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

This sentence doesn't make sense to me.
</blockquote></details>

Done.
___
*[pkg/sql/alter_column_type.go, line 276 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9P6w3k6WpCzxB33unG:-M9Q6PSTEM2ZDaP6l5xK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/alter_column_type.go#L276)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

This line is used to actually trigger the computed expression when a value is inserted into the new table. We need the computed expression to depend on the new column, I couldn't really think of a better way to do this besides using this sort of message.
</blockquote></details>

Done.
___
*[pkg/sql/alter_column_type.go, line 278 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9P70Wj0sIy_RJ9R7Ue:-M9Q6PyM4ClNKyw_WP2B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/alter_column_type.go#L278)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Could you parametrize '03000' here so it's not just a magic number? Also the use of `concat` here is kind of funky. I'd say do the string concatenation on the Go side.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/alter_column_type, line 484 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0EfArai3e557k0Au-r2-484:-M9Q6b5-4-q8J5YfGrIe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/020ba6a6ef37a6db94eefa1502a813eeb1ba40f2/pkg/sql/logictest/testdata/logic_test/alter_column_type#L484)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

It's good to keep the comments around 80 characters for readability still, even if the errors go larger than the limit.
</blockquote></details>

Done.
___
*[pkg/sql/schemaexpr/column.go, line 246 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9PAT6P3fsCsXW2Wguw:-M9Q6c1740mypEj2mWQg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/schemaexpr/column.go#L246)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Could you use `pgerror.Newf` with an appropriate `pgcode`? Also having a newline in an error msg seems weird to me.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rohany commented on pkg/sql/schemaexpr/column.go:
```diff
@@ -234,6 +234,27 @@ func replaceVars(
+			if c.TableName.NumParts > 1 {
+				return false, nil,
+					pgerror.Newf(pgcode.InvalidColumnReference,
+						"cannot reference columns with a specified database or schema")
```
I think something like "cannot qualify columns in expression with database or schema" would be clearer.

rohany commented on pkg/sql/schemaexpr/column.go:
```diff
@@ -234,6 +234,27 @@ func replaceVars(
+			if name.Table() != desc.Name {
+				return false, nil,
+					pgerror.Newf(pgcode.InvalidColumnReference,
+						"cannot reference columns from other tables, "+
```
Just "cannot reference column in table %s" would be fine here

rohany commented on pkg/sql/sqlbase/structured.proto:
```diff
@@ -536,6 +536,11 @@ message ComputedColumnSwap {
   option (gogoproto.equal) = true;
   optional uint32 new_column_id = 1 [(gogoproto.nullable) = false, (gogoproto.casttype) = "ColumnID"];
   optional uint32 old_column_id = 2 [(gogoproto.nullable) = false, (gogoproto.casttype) = "ColumnID"];
+  // inverse_expr represents the computed expression for the old column to
```
This comment doesn't make that much sense. Something like 

"InverseExpr is the expression to compute values for the old column once it is swapped for the new column" is simpler.

rohany commented on pkg/sql/schemaexpr/partial_index_test.go:
```diff
@@ -77,10 +77,12 @@ func TestIndexPredicateValidator_Validate(t *testing.T) {
-		{"bar.a AND foo.bar.b = 0", true, "a AND (b = 0)"},
+
+		//  Cannot use reference columns with database or schema names.
+		{"foo.bar.a", false, ""},
```
cc @mgartner is this needed behavior for partial indexes?

rohany reviewed (approved):
LGTM aside from some comment nits

solongordon reviewed (changes requested):
Also LGTM in general though I want to understand why it's necessary to include the inserted value in the `force_error` message.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-:-M9T_jZV8Cn_4pw7eMp-:botjatg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, @rohany, and @solongordon)

___
*[pkg/sql/alter_column_type.go, line 231 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0Ef9V4B7KoTu-rvr-r3-231:-M9T_IB9-B6v1ZTV9PC0:baua9wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f413fe01b2b91a8abc3720ddd2fd347f38658f/pkg/sql/alter_column_type.go#L231)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Doesn't seem like anything changed?
___
*[pkg/sql/alter_column_type.go, line 276 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9P6w3k6WpCzxB33unG:-M9TZFer-LkG4mxRf4-B:b-jzuaf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/alter_column_type.go#L276)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Could you explain more what goes wrong if you don't include the inserted value in the message? It's not a big deal but I want to understand.

Also could you take a second look at the message? There is some strange comma usage and spacing.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, @rohany, and @solongordon)

___
*[pkg/sql/alter_column_type.go, line 231 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0Ef9V4B7KoTu-rvr-r3-231:-M9TfVgBDr7qYirasf8V:bmgk5s7) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f413fe01b2b91a8abc3720ddd2fd347f38658f/pkg/sql/alter_column_type.go#L231)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Doesn't seem like anything changed?
</blockquote></details>

Hmm super weird, it doesn't show up as changed on reviewable but if you check Github it should be updated.
___
*[pkg/sql/alter_column_type.go, line 276 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9P6w3k6WpCzxB33unG:-M9TfhcsB1nJXyg0QCUr:bbpzmhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/alter_column_type.go#L276)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Could you explain more what goes wrong if you don't include the inserted value in the message? It's not a big deal but I want to understand.

Also could you take a second look at the message? There is some strange comma usage and spacing.
</blockquote></details>

It seems like reviewable doesn't have the updated version, can you check the diff using Github instead, maybe I can force a refresh on reviewable.

The reason we want to include the inserted message is because after the column swap, we have this period of time where not every node is on the version with the new column, therefore reads and writes can still happen to the old column. If USING EXPRESSION is provided, we don't know what to insert into the old column if a write happens to the new column, for the case were no expression is provided, we can use the opposite cast to insert values into the old column, ie if we alter type from int -> string, during the short period where after we swap the columns and the old column is not dropped yet, every insert into the new column will be tried to be inserted into the old column with a cast from string -> int. However we can't generally invert a user's given expression so we want to turn the column into read only during this time, both the old and new column. This is why the inserted value is needed, the inserted value is how we make the old column computed based on the new column, so when an insert happens into the new column, the force_error will trigger due to the computed expression, without it inserts the error wouldn't actually trigger when we insert into the new column, we want the new column to become readonly until the old column is dropped.

Also here is the section RFC highlighting the problem, https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20200401_alter_column_type.md#mixed-table-version-clusters


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, @rohany, and @solongordon)

___
*[pkg/sql/alter_column_type.go, line 231 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0Ef9V4B7KoTu-rvr-r3-231:-M9ThlAB7JuOJK13AwoF:b-z7gnp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f413fe01b2b91a8abc3720ddd2fd347f38658f/pkg/sql/alter_column_type.go#L231)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Hmm super weird, it doesn't show up as changed on reviewable but if you check Github it should be updated.
</blockquote></details>

Wait actually I think it was changed, I think Rohan was just referring to the "expression" that was capitalized, do you mean uncapitalized USING EXPRESSION?


<!-- Sent from Reviewable.io -->

solongordon reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, and @rohany)

___
*[pkg/sql/alter_column_type.go, line 231 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0Ef9V4B7KoTu-rvr-r3-231:-M9TvIIi7K0ogrEPrIhM:bn18vvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f413fe01b2b91a8abc3720ddd2fd347f38658f/pkg/sql/alter_column_type.go#L231)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Wait actually I think it was changed, I think Rohan was just referring to the "expression" that was capitalized, do you mean uncapitalized USING EXPRESSION?
</blockquote></details>

I see, looks fine. (It seems a little weird to me for it to be capitalized in USING EXPRESSION too, but we do that consistently so I'm ok with it.)
___
*[pkg/sql/alter_column_type.go, line 276 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9P6w3k6WpCzxB33unG:-M9TvYRg3rOtxx4y1SWz:buk88xs) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/alter_column_type.go#L276)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

It seems like reviewable doesn't have the updated version, can you check the diff using Github instead, maybe I can force a refresh on reviewable.

The reason we want to include the inserted message is because after the column swap, we have this period of time where not every node is on the version with the new column, therefore reads and writes can still happen to the old column. If USING EXPRESSION is provided, we don't know what to insert into the old column if a write happens to the new column, for the case were no expression is provided, we can use the opposite cast to insert values into the old column, ie if we alter type from int -> string, during the short period where after we swap the columns and the old column is not dropped yet, every insert into the new column will be tried to be inserted into the old column with a cast from string -> int. However we can't generally invert a user's given expression so we want to turn the column into read only during this time, both the old and new column. This is why the inserted value is needed, the inserted value is how we make the old column computed based on the new column, so when an insert happens into the new column, the force_error will trigger due to the computed expression, without it inserts the error wouldn't actually trigger when we insert into the new column, we want the new column to become readonly until the old column is dropped.

Also here is the section RFC highlighting the problem, https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20200401_alter_column_type.md#mixed-table-version-clusters
</blockquote></details>

Thanks, I was missing that computed expression wouldn't be invoked unless it explicitly depends on the inserted value. I also see now why you needed the SQL `concat` to include that value.

The message still looks weird to me, even in the Github diff, for instance `tried to insert , %s,  into %s`. Maybe things got screwed up when you tried to remove the concat? Feel free to bring it back.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, @rohany, and @solongordon)

___
*[pkg/sql/alter_column_type.go, line 276 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9P6w3k6WpCzxB33unG:-M9U1PtB9mf6gQZZ-mkP:bv4pctp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd592c76d38073b6e4b77cbb5fc054482a70b/pkg/sql/alter_column_type.go#L276)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Thanks, I was missing that computed expression wouldn't be invoked unless it explicitly depends on the inserted value. I also see now why you needed the SQL `concat` to include that value.

The message still looks weird to me, even in the Github diff, for instance `tried to insert , %s,  into %s`. Maybe things got screwed up when you tried to remove the concat? Feel free to bring it back.
</blockquote></details>

Yeah I see it, I brought back concat, it should be fixed now.
___
*[pkg/sql/schemaexpr/column.go, line 247 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9Ky41hPWXIvh_7aKiD-r5-247:-M9U68gJ-dlzVGilizlS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b72dd067d55817ace82d6258986e3619f6c8205/pkg/sql/schemaexpr/column.go#L247)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I think something like "cannot qualify columns in expression with database or schema" would be clearer.
</blockquote></details>

Done.
___
*[pkg/sql/schemaexpr/column.go, line 253 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9Ky41hPWXIvh_7aKiD-r5-253:-M9U695o3MtxLCglT09M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b72dd067d55817ace82d6258986e3619f6c8205/pkg/sql/schemaexpr/column.go#L253)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Just "cannot reference column in table %s" would be fine here
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.proto, line 539 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M7y0EfArai3e557k0B2-r5-539:-M9U6APA1pfqE3aUsKNm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b72dd067d55817ace82d6258986e3619f6c8205/pkg/sql/sqlbase/structured.proto#L539)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

This comment doesn't make that much sense. Something like 

"InverseExpr is the expression to compute values for the old column once it is swapped for the new column" is simpler.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @mgartner, @RichardJCai, @rohany, and @solongordon)

___
*[pkg/sql/logictest/testdata/logic_test/partial_index, line 110 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9VAS4s1eTeryyaOXSf:-M9VAS4s1eTeryyaOXSg:b-uiq8fg) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd08ff1c66262717e2e1dcbef6646a7e18204e9b/pkg/sql/logictest/testdata/logic_test/partial_index#L110)):*
> ```
> CREATE INDEX t6i1 ON t6 (a) WHERE a > 6;
> CREATE INDEX t6i2 ON t6 (a) WHERE t6.a > 7;
> CREATE INDEX t6i3 ON t6 (a DESC) WHERE test.t6.a > 8;
> ```

This is a case we have to support in order to be compatible with Postgres.

In Postgres:

```
marcus=# create table t (i int);
CREATE TABLE

marcus=# create index idx on t (i) where public.t.i = 1;
CREATE INDEX

marcus=# \d t
                 Table "public.t"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 i      | integer |           |          |
Indexes:
    "idx" btree (i) WHERE i = 1
```

You mentioned that database/schema names are normalized and I agree that if they are normalized before checking if they are valid, that would lead to a bug. But I don't see any current bug in CRDB:

```
root@127.0.0.1:52830/defaultdb> set experimental_partial_indexes=on;
SET

Time: 250µs

root@127.0.0.1:52830/defaultdb> create table t (i int);
CREATE TABLE

Time: 3.104ms

root@127.0.0.1:52830/defaultdb> create index idx on t (i) where badschema.t.i = 1;
ERROR: no data source matches prefix: badschema.t
SQLSTATE: 42P01
```

Am I missing something?
___
*[pkg/sql/schemaexpr/partial_index_test.go, line 83 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9Q82wMrBQMEMTkV3rJ-r5-83:-M9VCMxC6jAS32vhqGPY:bowrvzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b72dd067d55817ace82d6258986e3619f6c8205/pkg/sql/schemaexpr/partial_index_test.go#L83)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

cc @mgartner is this needed behavior for partial indexes?
</blockquote></details>

Yes. See my other comment.


<!-- Sent from Reviewable.io -->

RichardJCai committed:
sql: Support ALTER COLUMN TYPE with a user provided expression.

Fixes #47706

Release note (sql change): Support ALTER COLUMN TYPE ... USING EXPRESSION
to allow users to optionally provide an expression to convert a column to a new type.

mgartner reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 11 files at r6, 4 of 4 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-:-M9VWoFd0HrqGfNE6B99:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, @rohany, and @solongordon)

___
*[pkg/sql/alter_column_type.go, line 236 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9VTPYF6rieze8Y376_:-M9VTPYG5QGcgDAQHd8k:b-hceqrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/04fd3a705711658d730e0a452ab2f043c40154dc/pkg/sql/alter_column_type.go#L236)):*
> ```Go
> 	// will fail until the old column is dropped.
> 	var inverseExpr string
> 	if Using != nil {
> ```

[nit] Why is `Using` capitalized?
___
*[pkg/sql/logictest/testdata/logic_test/alter_column_type, line 462 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9VTr5-9_seyp9mIxwZ:-M9VTr5-9_seyp9mIxw_:b-ktg9cc) ([raw file](https://github.com/cockroachdb/cockroach/blob/04fd3a705711658d730e0a452ab2f043c40154dc/pkg/sql/logictest/testdata/logic_test/alter_column_type#L462)):*
> ```
> true
> 
> # Ensure ALTER COLUMN TYPE rolls back if  is not applicable to value in the column.
> ```

[nit] extra space between "if" and "is"
___
*[pkg/sql/logictest/testdata/logic_test/alter_column_type, line 495 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9VUKEKB-dth7S9TwmB:-M9VUKEKB-dth7S9TwmC:b80y5i1) ([raw file](https://github.com/cockroachdb/cockroach/blob/04fd3a705711658d730e0a452ab2f043c40154dc/pkg/sql/logictest/testdata/logic_test/alter_column_type#L495)):*
> ```
> # table being altered.
> statement error pq: column "y" does not exist
> ALTER TABLE t26 ALTER COLUMN x TYPE BOOL USING(y > 0);
> ```

[nit] add a space after `USING` here and in the commented tests below
___
*[pkg/sql/schemaexpr/column.go, line 256 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9VUl3A1BQohnyk1C3J:-M9VUl3A1BQohnyk1C3K:b-l78i9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/04fd3a705711658d730e0a452ab2f043c40154dc/pkg/sql/schemaexpr/column.go#L256)):*
> ```Go
> // ReplaceColumnVarsAndSanitizeExpr takes an expr and replaces all column
> // variables with a dummyColumn of the column's type and then sanitizes
> // the expression.
> ```

[nit] maybe explain what "sanitizes the expression" means. up to you.
___
*[pkg/sql/schemaexpr/column.go, line 257 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9VWFpy5PEjz_xTYxaV:-M9VWFpy5PEjz_xTYxaW:b-datmvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/04fd3a705711658d730e0a452ab2f043c40154dc/pkg/sql/schemaexpr/column.go#L257)):*
> ```Go
> // variables with a dummyColumn of the column's type and then sanitizes
> // the expression.
> func ReplaceColumnVarsAndSanitizeExpr(
> ```

Not necessary for you to do anything now, but I wanted to mention that maybe this function (and some other functions we've already added to `schemaexpr/column.go`) don't really fit in with the file name. Maybe some of these belong in a new `schemaexpr/expr.go` file. I'd also like to find a way to clean up some of these function signatures have 4+ arguments. I'll noodle on it a bit.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @mgartner, @rohany, and @solongordon)

___
*[pkg/sql/alter_column_type.go, line 236 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9VTPYF6rieze8Y376_:-M9ZB4xF2Uc4S6F60FQg:bw1h10o) ([raw file](https://github.com/cockroachdb/cockroach/blob/04fd3a705711658d730e0a452ab2f043c40154dc/pkg/sql/alter_column_type.go#L236)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

[nit] Why is `Using` capitalized?
</blockquote></details>

No idea :P changed it to lower case.
___
*[pkg/sql/logictest/testdata/logic_test/partial_index, line 110 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9VAS4s1eTeryyaOXSf:-M9ZBB2r-qFNlzAv5UKL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd08ff1c66262717e2e1dcbef6646a7e18204e9b/pkg/sql/logictest/testdata/logic_test/partial_index#L110)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This is a case we have to support in order to be compatible with Postgres.

In Postgres:

```
marcus=# create table t (i int);
CREATE TABLE

marcus=# create index idx on t (i) where public.t.i = 1;
CREATE INDEX

marcus=# \d t
                 Table "public.t"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 i      | integer |           |          |
Indexes:
    "idx" btree (i) WHERE i = 1
```

You mentioned that database/schema names are normalized and I agree that if they are normalized before checking if they are valid, that would lead to a bug. But I don't see any current bug in CRDB:

```
root@127.0.0.1:52830/defaultdb> set experimental_partial_indexes=on;
SET

Time: 250µs

root@127.0.0.1:52830/defaultdb> create table t (i int);
CREATE TABLE

Time: 3.104ms

root@127.0.0.1:52830/defaultdb> create index idx on t (i) where badschema.t.i = 1;
ERROR: no data source matches prefix: badschema.t
SQLSTATE: 42P01
```

Am I missing something?
</blockquote></details>

Done.
___
*[pkg/sql/schemaexpr/column.go, line 257 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9VWFpy5PEjz_xTYxaV:-M9ZBCGC8Qm5ogpDzSP3:b-t6nkyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/04fd3a705711658d730e0a452ab2f043c40154dc/pkg/sql/schemaexpr/column.go#L257)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Not necessary for you to do anything now, but I wanted to mention that maybe this function (and some other functions we've already added to `schemaexpr/column.go`) don't really fit in with the file name. Maybe some of these belong in a new `schemaexpr/expr.go` file. I'd also like to find a way to clean up some of these function signatures have 4+ arguments. I'll noodle on it a bit.
</blockquote></details>

Yeah I noticed that we have a lot of these functions that take a lot of arguments as well, I generally dislike having a lot of function arguments but I haven't had any gripes with this so far, for some reason having a lot of function args in go seems manageable still, what are your thoughts?
___
*[pkg/sql/schemaexpr/partial_index_test.go, line 83 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9Q82wMrBQMEMTkV3rJ-r5-83:-M9ZC4Yx9wthfGQ_mcbv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b72dd067d55817ace82d6258986e3619f6c8205/pkg/sql/schemaexpr/partial_index_test.go#L83)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Yes. See my other comment.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RichardJCai commented:
Any more comments on this before merging?

mgartner reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49452)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @mgartner, @rohany, and @solongordon)

___
*[pkg/sql/schemaexpr/column.go, line 257 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49452#-M9VWFpy5PEjz_xTYxaV:-M9_SWQF37P5KIFWZpv3:b-8squv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/04fd3a705711658d730e0a452ab2f043c40154dc/pkg/sql/schemaexpr/column.go#L257)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Yeah I noticed that we have a lot of these functions that take a lot of arguments as well, I generally dislike having a lot of function arguments but I haven't had any gripes with this so far, for some reason having a lot of function args in go seems manageable still, what are your thoughts?
</blockquote></details>

Leave as-is for now :+1: 


<!-- Sent from Reviewable.io -->

RichardJCai commented:
bors r+

RichardJCai commented:
Thanks for the reviews all!
